refactor apikey related endpoints and data model

wenyang-cao opened the pull request:
This PR is for code review only. 

1.Data Model Updates:
The API key table now uses UUID as the foreign key instead of username.
Legacy columns of type String have been updated to use more standardized and appropriate types.

2.Query Logic:

Due to the foreign key switch , some data access patterns have changed accordingly.

3. Refactoring: This PR incorporates changes previously requested via code review comments.

While the overall scope is not large, the changes are notable so another review of all related code is necessary.

Testing: All unit tests pass.

This PR does not include any changes to the authentication logic.

kroczi commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,274 @@
+    } yield (user, key)
+
+    val fullQuery = for {
+      resultOpt   <- userAndKeyQuery.result.headOption
+      ownerStrOpt <- resultOpt match {
+                        case Some(_) => ApiKeysTQ.getOrgAndUsernameByKeyId(keyid)
+                        case None    => DBIO.successful(None)
+                      }
```
You should already have owner id from the `resultOpt` - no need to fetch the the API key once more in `getOrgAndUsernameByKeyId`.

kroczi commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,274 @@
+def userApiKeys(identity: Identity2): Route = {
+  pathPrefix("orgs" / Segment / "users" / Segment / "apikeys") { (organization, username) =>
+    val resource = OrgAndId(organization, username).toString
+    val resourceType = "user"
+
```
not used

kroczi commented on src/main/scala/org/openhorizon/exchangeapi/route/user/User.scala:
```diff
@@ -130,23 +139,50 @@ trait User extends JacksonSupport with AuthenticationSupport {
+                                                new ApiKeyMetadata(apiKeyRow, null)
+                                          )
+
+                                          val user = new UserTable(userResult, Some(apiKeyMetadataList))
+                                          val userMap: Map[String, UserTable] =
+                                                Map(s"${userRow.organization}/${userRow.username}" -> user)
+
+                                          (StatusCodes.OK, GetUsersResponse(userMap, 0))
```
This part is repeated, can be extracted after `db.run()`.

kroczi commented on src/main/scala/org/openhorizon/exchangeapi/route/user/Users.scala:
```diff
@@ -133,23 +149,58 @@ trait Users extends JacksonSupport with AuthenticationSupport {
+                  (StatusCodes.OK, GetUsersResponse(userMap, 0))
+
+                case Failure(t) =>
+                  logger.error(s"Failed to fetch API keys for users: ${t.getMessage}")
+
```
If none of the selected users has API key in the DB - will we get an error in the logs?

wenyang-cao replied:
No, this error log will not be triggered when a user simply has no API keys. In that case, the query result just returns an empty sequence, which is still a Success.  It  only runs if the DB query itself fails and returns a Failure

kroczi commented on src/main/scala/org/openhorizon/exchangeapi/route/user/Users.scala:
```diff
@@ -133,23 +149,58 @@ trait Users extends JacksonSupport with AuthenticationSupport {
+                case Failure(t) =>
+                  logger.error(s"Failed to fetch API keys for users: ${t.getMessage}")
+
+                  val userMap: Map[String, User] = result.map { userResult =>
+                    val userRow = userResult._1
+                    val user = new User(userResult, Some(Seq.empty))
+                    s"${userRow.organization}/${userRow.username}" -> user
+                  }.toMap
+
+                  (StatusCodes.OK, GetUsersResponse(userMap, 0))
```
This fragment is repeated - I guess it can be extracted outside `db.run()`

kroczi commented on src/main/scala/org/openhorizon/exchangeapi/route/user/Users.scala:
```diff
@@ -133,23 +149,58 @@ trait Users extends JacksonSupport with AuthenticationSupport {
+            } else {
+              val userUuids = result.map(_._1.user)
+              val allApiKeysQuery = ApiKeysTQ.filter(_.user.inSet(userUuids))
+
+              db.run(allApiKeysQuery.result.asTry).map {
```
There is only one db query - do we need to wrap it in the Future?

wenyang-cao replied:
We could remove .asTry and just use .map + .recover instead of destructuring Try. I can clean that up.

naphelps replied:
All db queries should (implicit or not) get wrapped in a Future. Welcome to the syntactical sugar games that is Scala. If the db query were not appropriately wrapped you would have to use `Await.[result/ready](db.run(...), 15.seconds)`which is blocking IO. The `complete` above wraps a Future as part of the directive. You still could, if it made sense, nest the thread forks.

wenyang-cao commented on src/main/resources/messages.txt:
```diff
@@ -47,6 +47,10 @@ api.internal.error=internal_error
 api.invalid.input=invalid-input
 api.not.found=not-found
 api.not.implemented=not-implemented
+apikey.creation.failed=Failed to create API key.
+apikey.deletion.failed=An internal error occurred when deleting the API key
+apikey.deleted=API key deleted successful
+apikey.not.found=API key not found
 arch.cannot.be.empty=arch can not be empty.
 attribute.not.part.of.service=Service attribute name ''{0}'' is not an attribute of the service resource.
 bad.auth.header=Basic auth header is missing '':'' or is bad encoded format
```
Do you mean a dot at the end of the sentence? I just double check this file. The existing message file already has mixed styles. most messages don’t have a period, but a few do. I’ll remove the "." from the message I added.

If we want to standardize the rest, it should be handled separately.

kroczi commented:
@wenyang-cao @naphelps don't we need to generate openapi json schema for created/updated endpoints?

wenyang-cao commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,249 @@
+
+          case Success(Left(msg)) =>
+            (HttpCode.BAD_INPUT, ApiResponse(ApiRespType.BAD_INPUT, msg))
+          case Failure(_) =>
+            (HttpCode.BAD_INPUT, ApiResponse(ApiRespType.BAD_INPUT, ExchMsg.translate("apikey.creation.failed")))
```
Currently we don't have any validation logic for the description field. So I think BadInput only applies to malformed request bodies and cases where the user doesn't exist. The latter is already handled explicitly in case Success(Left(msg)).

Since malformed request bodies would likely be rejected earlier by the framework, any failure at this point would be unexpected. Given that, it makes sense to return InternalServerError here. I'll switch the http code accordingly.

wenyang-cao commented on src/main/scala/org/openhorizon/exchangeapi/table/user/User.scala:
```diff
@@ -29,4 +31,21 @@ final case class User(admin: Boolean = false,
              case _ =>
                ""
            })
+  def this(tuple: (UserRow, Option[(String, UUID, String)]), apikeys: Option[Seq[ApiKeyMetadata]]) =
```
When I first added apikeys, I kept both constructors because the old code used the one (new User(...))  to construct a user in a few places. Removing it would have required explicitly passing None, so I left it to avoid unnecessary changes.

Now that I’ve checked again, it looks like only my new code uses new User(...), and only with an actual apikeys value. Everywhere else just uses the main constructor with default params.

So unless we expect new User(...) to be used again without apikeys, it should be safe to remove the old constructor. Let me know if you're good with that and I’ll clean it up.

naphelps commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,274 @@
+        mediaType = "application/json", 
+        schema = new Schema(implementation = classOf[PostApiKeyResponse]),
+        examples = Array(
+          new ExampleObject(value = """{
```
These examples use pre-formatted text `""" ... """` denotes this. keep in mind the indentation/whitespace you add here will show in the rendered swagger doc. You will want to construct the examples off of the left edge.

naphelps commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,274 @@
+      val timestamp: java.sql.Timestamp = ApiTime.nowUTCTimestamp //standardizing on Timestamp DB types
+
+      complete {
+      db.run((for {
```
This needs some rework. The nested for comprehensions are redundant,everything can be done in a single structure. val row can be added directly at the insert, the val is wasted. You mixed syntax/styles on the inserts between the apikey table and the changes table. preference `+=` and `++=` where and when needed. The `.to..Row` function is a waste,build the "row" object directly. The user table query at the beginning does not have a purpose. If you need the UUID its in the caller identity that is pasted by the authenticator at the top of the route stack,`Identity2.identifier` for user archetypes. If the Insert bounces off the database because the user does not exist. This is a good thing. Results in an automatic rollback,and we can catch the Exception and return the appropriate http code and msg further down in the `.map {...}` after the DB call.

wenyang-cao replied:
@naphelps Thanks for the feedback. I'll make changes accordingly. However, I believe the part about removing the user query (UsersTQ.filter(...)) deserves a second look.

While identity.identifier.get gives us the authenticated caller's UUID, the API key being created may be for a different user. i.e., an org admin can create an API key for someone else. In that case, we can't rely on the caller's identity — we need the UUID of the actual user identified by the path params (organization, username).

That said, I think in this case I can simplify the query to just extract the UUID (.map(_.user).result.head) instead of fetching the entire user row.

naphelps replied:
Ah, yes you are right. Good thing there is one for the callee as well! Even easier its cached (`cacheResourceOwnership`). We can also do checks to skip any unnecessary DB steps if say `resource != identity.resource`.

naphelps commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,274 @@
+      mediaType = "application/json",
+      schema = new Schema(implementation = classOf[PostApiKeyRequest]),
+      examples = Array(
+        new ExampleObject(value = """{
```
Be mindful of the pre-formatted text.

naphelps commented on src/main/scala/org/openhorizon/exchangeapi/utility/ApiKeyUtils.scala:
```diff
@@ -0,0 +1,26 @@
+package org.openhorizon.exchangeapi.utility
```
I would stick all of the Crypto things up in the auth/ package.

naphelps replied:
I would still prefer moving it.

wenyang-cao commented:
Updated based on review feedback. Have rebased onto the latest state of the base branch.

- Cleaned up indentation, formatting, and Swagger annotations.

- Removed redundant or unused code, and simplified logic where possible.

- Replaced custom hashing with Password.hash() (Argon2).

- For user routes, removed use of .asTry and replaced destructuring of Try with .map + .recover.

- Refactored the Post API key route to flatten DB logic and align with preferred insert style.

- Refactored the Get by ID API key route by removing getOrgAndUsernameByKeyId and get ownerstr directly via query result.

@kroczi  @naphelps  Please take a look when you have time. Thanks

naphelps commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,308 @@
+  def postUserApiKey(@Parameter(hidden = true) identity: Identity2,
+                     @Parameter(hidden = true) organization: String,
+                     @Parameter(hidden = true) username: String,
+                     resourceUuidOpt: Option[UUID]): Route = {
```
missing hiding parameter.

naphelps commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,308 @@
+                       @Parameter(hidden = true) organization: String,
+                       @Parameter(hidden = true) username: String,
+                       @Parameter(hidden = true) keyid: UUID,
+                       resourceUuidOpt: Option[UUID]): Route = complete {
```
missing hiding parameter.

naphelps commented on src/main/scala/org/openhorizon/exchangeapi/route/user/Users.scala:
```diff
@@ -133,24 +149,54 @@ trait Users extends JacksonSupport with AuthenticationSupport {
+                        val userUuids = result.map(_._1.user)
+                        val allApiKeysQuery = ApiKeysTQ.filter(_.user.inSet(userUuids))
+
+                        db.run(allApiKeysQuery.result).map { allApiKeys =>
```
The trick of flatmaps on threads is they chain.

kroczi commented on src/main/scala/org/openhorizon/exchangeapi/route/apikey/UserApiKeys.scala:
```diff
@@ -0,0 +1,332 @@
+    resourceUuidOpt match {
+      case Some(userUuid) =>
+        db.run((for {
+          deleted <- ApiKeysTQ.filter(k => k.id === keyid && k.user === userUuid && k.orgid === organization)
+                             .filterIf(identity.isStandardUser)(k => k.user === identity.identifier.get)
+                             .filterIf(identity.isOrgAdmin)(k => k.orgid === identity.organization)
+                             .filterIf(identity.isHubAdmin)(k => UsersTQ.filter(u => u.user === k.user && ((u.isHubAdmin || u.isOrgAdmin) && !(u.organization === "root" && u.username === "root"))).exists)
```
I guess the `filterIf`s are only needed if the identity != (organization, username),

kroczi replied:
Also will this work for `root` user if he wants to alter his own API keys?
`root` user is HubAdmin and OrgAdmin (in `root` org) which means the last `filterIf` will activate and it will filter out his keys based on the last clause?

```
$ hzn ex user ls -o root
{
  "root/root": {
    "password": "",
    "admin": true,
    "hubAdmin": true,
    "lastUpdated": "2025-06-07T20:07:34.050372442Z[UTC]",
    "updatedBy": ""
  }
}
```

kroczi replied:
Please add a test case where root user alters his own API key.

kroczi replied:
But then still the `root/root` is like a super user - should he be able to modify any user in the system? right now he will only be able to modify only other admins.

kroczi replied:
Even worse. He will be only able to modify other admins in his org (`root`) (that are not himself) due to activation of both second and third filter.

Any user that is both hubadmin and orgadmin will have such narrowed power. (Because he will be granted power over intersection, not union of his two roles).

Please add a test case for it.

wenyang-cao replied:
@kroczi I believe Hub Admin users must be created in the root org so the root user will be the only one who can be both orgadmin and hubadmin. And like Nathan has just said there is no filter for root user so it can see everything and do anything.

kroczi replied:
@wenyang-cao @naphelps Thanks for explanation. It looks good then.

Sorry for the bunch of stupid questions.

wenyang-cao commented:
@kroczi I‘ve added test cases for root also. The test cases verify that root can post, get, and delete apikeys for themselves, normal org admins, hub admins. I hope this help address any remaining concerns with the permission of root user.

kroczi commented:
Changes look good to me. Thanks @wenyang-cao for addressing the comments.
